tortoise/backends/base/client.py

```diff
@@ -7,6 +7,18 @@
 from tortoise.backends.base.schema_generator import BaseSchemaGenerator
 
 
+class Capabilities:
+    def __init__(self, dialect: str, *, connection: dict = {}, **kwargs) -> None:
+        self.dialect = dialect
+        self.connection = connection
```

Comment:
Why does capability need connections params?

Also don't think using mutable collections as default value is good. 
Better be safe and set it to `None` and create new dict inside `__init__`:
`self.connection = connection or {}`

Reply:
I treated the Capability object as a config object. Not sure if that is wise or not.
Now that I think of it, it should in-fact be a a read-only informational property.
And yes, mutable defaults are bad. In fact I think Capability should be immutable by default.

Reply:
Now that I think of it, the reason is that capabilites can change depending on the connection itself. e.g. we need >= <some-version> of a DB to get the capabilities. or the GIS plugin needs to be installed?

Reply:
+1 about the `{}` default, mutable defaults cause all sorts of weird things in Python.

Reply:
I changed `Capabilities` to not leak changes made in there back to the DB Client.
Then went a step further and tried to make it behave somewhat immutable (as much as Python allows one to make things immutable). This conflicted with typing and autocompletion, hence my weird `_mutable` thing there.

Reply:
NamedTuples are unfortunately more trouble than they are worth. (https://rhye.org/post/python-cassandra-namedtuple-performance/, https://www.attrs.org/en/stable/why.html#namedtuples)
They are immutable, but their content is still references, meaning if you put a mutable type inside it (e.g. dict) then it would allow you to alter state. So to try and prevent accidentally changing it, we would have to pass in deepcopies of everything manually.

I wanted to use attrs, but feel it unfair to introduce as a dependency for just this use case.

Data Classes would also work, but is only part of Py3.7 and backportable to Py3.6, which unfortunately makes them a non-option.

So I settled on a class.

Also, in this case explicit isn't better, as we want to minimise breakage of external drivers.
We state that capabilities are pessimistic by default.

And Yes, since we only case about capabilities we explicity know about, I should probably drop the kwargs part.

Reply:
Understood! Makes sense.

Reply:
@abondar Any further comments?